src/jit/MemoryInl.h

```diff
@@ -215,11 +246,81 @@ void MemAddress::load(sljit_compiler* compiler)
 #endif /* HAS_SIMD */
 }
 
+#if defined(ENABLE_EXTENDED_FEATURES) && (defined SLJIT_32BIT_ARCHITECTURE && SLJIT_32BIT_ARCHITECTURE)
+static void atomicRmwLoad64(int64_t* shared_p, uint64_t* result)
+{
+    std::atomic<uint64_t>* shared = reinterpret_cast<std::atomic<uint64_t>*>(shared_p);
+    *result = shared->load(std::memory_order_relaxed);
+}
+
+static void atomicRmwStore64(uint64_t* shared_p, int64_t* value)
+{
+    std::atomic<uint64_t>* shared = reinterpret_cast<std::atomic<uint64_t>*>(shared_p);
+    shared->store(*value);
+}
+
+static void emitAtomicLoadStore64(sljit_compiler* compiler, Instruction* instr)
+{
+    uint32_t options = MemAddress::CheckNaturalAlignment | MemAddress::DontUseOffsetReg;
+    uint32_t size = 8;
+    sljit_u32 offset;
+
+    Operand* operands = instr->operands();
+    MemAddress addr(options, instr->requiredReg(0), instr->requiredReg(1), 0);
+
+    if (instr->opcode() == ByteCode::I64AtomicLoadOpcode) {
+        MemoryLoad* loadOperation = reinterpret_cast<MemoryLoad*>(instr->byteCode());
+        offset = loadOperation->offset();
+    } else {
+        ASSERT(instr->opcode() == ByteCode::I64AtomicStoreOpcode);
+        MemoryStore* storeOperation = reinterpret_cast<MemoryStore*>(instr->byteCode());
+        offset = storeOperation->offset();
+    }
+
+    addr.check(compiler, operands, offset, size);
+
+    if (addr.memArg.arg == 0) {
+        return;
+    }
+
+    JITArgPair valueArgPair(operands + 1);
+    sljit_s32 type = SLJIT_ARGS2V(P, P);
+    sljit_s32 faddr;
+    if (instr->opcode() == ByteCode::I64AtomicLoadOpcode) {
+        faddr = GET_FUNC_ADDR(sljit_sw, atomicRmwLoad64);
+    } else {
+        ASSERT(instr->opcode() == ByteCode::I64AtomicStoreOpcode);
+        faddr = GET_FUNC_ADDR(sljit_sw, atomicRmwStore64);
+    }
+
+    if (valueArgPair.arg1 != SLJIT_MEM1(kFrameReg)) {
```

Comment:
This should be done only for stores, am I right? Could be moved into the else above.